File: jun1/src/combinators/forward_ref.py
from typing import Optional, cast

from ..combinator import Combinator
from ..parse_iteration_result import FrameStack, ParserIterationResult
from ..state import CombinatorState


class ForwardRefState(CombinatorState):
    def __init__(self, inner_state: Optional[CombinatorState] = None) -> None:
        self.inner_state = inner_state


class ForwardRef(Combinator):
    def __init__(self, combinator: Optional[Combinator] = None) -> None:
        self.combinator = combinator

    def set(self, combinator: Combinator) -> None:
        self.combinator = combinator

    def initial_state(self, signal_id: int, frame_stack: FrameStack) -> ForwardRefState:
        if self.combinator is None:
            raise Exception("ForwardRef not set")
        return ForwardRefState(self.combinator.initial_state(signal_id, frame_stack))

    def next_state(self, state: ForwardRefState, c: Optional[str], signal_id: int) -> tuple[ForwardRefState, int, ParserIterationResult]:
        if state.inner_state is None or self.combinator is None:
            raise Exception("Forward reference not set before use")
        state.inner_state, signal_id, result = self.combinator.next_state(state.inner_state, c, signal_id)
        return state, signal_id, result

File: jun1/src/combinators/forward_ref.rs
use std::cell::RefCell;
use std::rc::Rc;
use crate::combinator::Combinator;
use crate::parse_iteration_result::{FrameStack, ParserIterationResult};
use crate::state::CombinatorState;

pub struct ForwardRef<C>(pub Rc<RefCell<Option<C>>>);

impl<C> Combinator for ForwardRef<C>
where
    C: Combinator<State = Box<dyn CombinatorState>> + 'static,
{
    type State = Box<ForwardRefState<Box<dyn CombinatorState>>>;

    fn initial_state(&self, signal_id: &mut usize, frame_stack: FrameStack) -> Self::State {
        Box::new(ForwardRefState {
            inner_state: self.0.borrow().as_ref().map(|c| c.initial_state(signal_id, frame_stack)),
        })
    }

    fn next_state(&self, state: &mut Self::State, c: Option<char>, signal_id: &mut usize) -> ParserIterationResult {
        match state.inner_state.as_mut() {
            Some(inner_state) => match self.0.borrow().as_ref() {
                Some(combinator) => combinator.next_state(inner_state, c, signal_id),
                None => panic!("Forward reference not set before use"),
            },
            None => panic!("Forward reference not set before use"),
        }
    }
}

pub struct ForwardRefState<State> {
    pub inner_state: Option<State>,
}

impl<State: CombinatorState + 'static> CombinatorState for ForwardRefState<State> {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}